haystack/preview/dataclasses/document.py

```diff
@@ -137,15 +137,28 @@ def from_dict(cls, data: Dict[str, Any]) -> "Document":
             data["dataframe"] = pandas.read_json(io.StringIO(dataframe))
         if blob := data.get("blob"):
             data["blob"] = ByteStream(data=bytes(blob["data"]), mime_type=blob["mime_type"])
-        # Unflatten metadata if it was flattened
-        meta = {}
+        # Store metadata for a moment while we try un-flattening allegedly flatten metadata.
+        # We don't expect both a `meta=` keyword and flatten metadata keys so we'll raise a
+        # ValueError later if this is the case.
+        meta = data.pop("meta", {})
+        # Unflatten metadata if it was flattened. We assume any keyword argument that's not
+        # a document field is a metadata key. We treat legacy fields as document fields
+        # for backward compatibility.
+        flatten_meta = {}
         legacy_fields = ["content_type", "id_hash_keys"]
-        field_names = legacy_fields + [f.name for f in fields(cls)]
+        document_fields = legacy_fields + [f.name for f in fields(cls)]
         for key in list(data.keys()):
-            if key not in field_names:
-                meta[key] = data.pop(key)
+            if key not in document_fields:
+                flatten_meta[key] = data.pop(key)
 
-        return cls(**data, meta=meta)
+        # We don't support passing both flatten keys and the `meta` keyword parameter
+        if meta and flatten_meta:
+            raise ValueError(
+                "Passing the 'meta' parameter together with flatten metadata keys as keyword arguments is not supported."
```

Comment:
To focus on what we do support :) how about:
You can pass either the `meta` parameter or flattened metadata keys as keyword arguments, but currently you're passing both. Pass either the `meta` parameter or flattened metadata keys.